worker/introspection/leases.go

```diff
@@ -187,3 +199,138 @@ func (h leaseHandler) format(leases *leases) map[string]interface{} {
+		return
+	}
+
+	snapshot, err := h.snapshot()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintln(w, err.Error())
+		return
+	}
+
+	holder, ok := snapshot.Entries[leaseKey]
+	if !ok {
+		w.WriteHeader(http.StatusBadRequest)
+		if leaseKey.Namespace == lease.SingularControllerNamespace {
+			fmt.Fprintf(w, "singular lease for model %q not found\n", leaseKey.ModelUUID)
+		} else {
+			fmt.Fprintf(w, "application lease for model %q and app %q not found\n", leaseKey.ModelUUID, leaseKey.Lease)
+		}
+		return
+	}
+
+	if err := h.revokeLeadership(leaseKey, holder.Holder); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintln(w, err.Error())
+		return
+	}
+
+	if leaseKey.Namespace == lease.SingularControllerNamespace {
+		fmt.Fprintf(w, "singular lease for model %q revoked\n", leaseKey.ModelUUID)
+	} else {
+		fmt.Fprintf(w, "application lease for model %q and app %q revoked\n", leaseKey.ModelUUID, leaseKey.Lease)
+	}
+}
+
+func (h *leaseHandler) revokeLeadership(key raftlease.SnapshotKey, holder string) error {
+	command := &raftlease.Command{
+		Version:   raftlease.CommandVersion,
+		Operation: raftlease.OperationRevoke,
+		Namespace: key.Namespace,
+		ModelUUID: key.ModelUUID,
+		Lease:     key.Lease,
+		Holder:    holder,
+	}
+
+	bytes, err := command.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if h.runID == 0 {
+		source := rand.NewSource(h.clock.Now().UnixNano())
+		h.runID = rand.New(source).Int31()
+	}
+
+	// TODO(wallyworld): this should really be in a shared messages.go file under
+	// the top level pubsub package.
+	leaseRequestTopic := "lease.request"
+	requestID := atomic.AddUint64(&h.requestID, 1)
+	responseTopic := fmt.Sprintf("%s.%08x.%d", leaseRequestTopic, h.runID, requestID)
+
+	responseChan := make(chan raftlease.ForwardResponse, 1)
+	errChan := make(chan error)
+	unsubscribe, err := h.hub.Subscribe(
+		responseTopic,
+		func(_ string, resp raftlease.ForwardResponse, err error) {
+			if err != nil {
+				errChan <- err
+				return
+			}
+			responseChan <- resp
+		},
+	)
+	if err != nil {
+		return errors.Annotatef(err, "running %s", command)
+	}
+	defer unsubscribe()
+
+	_, err = h.hub.Publish(leaseRequestTopic, raftlease.ForwardRequest{
+		Command:       string(bytes),
+		ResponseTopic: responseTopic,
+	})
+	if err != nil {
+		return errors.Annotatef(err, "publishing %s", command)
+	}
+
+	select {
+	case <-h.clock.After(15 * time.Second):
+		return lease.ErrTimeout
+	case err := <-errChan:
+		return errors.Trace(err)
+	case response := <-responseChan:
+		return raftlease.RecoverError(response.Error)
+	}
+}
+
+func (h *leaseHandler) parseRevokeForm(r *http.Request) (raftlease.SnapshotKey, error) {
+	var result raftlease.SnapshotKey
+	if err := r.ParseForm(); err != nil {
+		return result, errors.Annotate(err, "parse form")
+	}
+
+	result.ModelUUID = r.Form.Get("model")
+	if result.ModelUUID == "" {
+		return result, errors.New("missing model")
```

Comment:
Missing model or model uuid?

Reply:
we tend to use them interchangeably. i added uuid to be explicit